File: 0211_design_add_and_search_words_data_structure.swift
// https://leetcode.com/problems/design-add-and-search-words-data-structure/

class WordDictionary {
	private class WordNode {
		var dic = Dictionary<Character, WordNode>()
	}

	private var wordRoot: WordNode

	init() {
		wordRoot = WordNode()
	}

	func addWord(_ word: String) {
		let word = word + "/"
		var node = wordRoot
		for char in word {
			if let next = node.dic[char] {
				node = next
			} else {
				let next = WordNode()
				node.dic[char] = next
				node = next
			}
		}
	}

	func search(_ word: String) -> Bool {
		let word = word + "/"
		var nodes = [wordRoot]

		for char in word {
			var newNodes = [WordNode]()
			if char == "." {
				for node in nodes {
					for value in node.dic.values {
						newNodes.append(value)
					}
				}
			} else {
				for node in nodes {
					if let next = node.dic[char] {
						newNodes.append(next)
					}
				}
			}

			if newNodes.count == 0 { return false }
			nodes = newNodes
		}

		return true
	}
}

File: 0211_design_add_and_search_words_data_structure.kt
// https://leetcode.com/problems/design-add-and-search-words-data-structure/description/

class WordDictionary() {
    private val root = TrieNode()

    fun addWord(word: String) {
        val wordWithEnd = "$word#"
        var node = root
        for (c in wordWithEnd) {
            node.nexts[c]?.let {
                node = it
            } ?: run {
                val newNode = TrieNode()
                node.nexts[c] = newNode
                node = newNode
            }
        }
    }

    fun search(word: String): Boolean {
        val wordWithEnd = "$word#"
        var nodes = arrayListOf(root)
        for (c in wordWithEnd) {
            val newNodes = ArrayList<TrieNode>()
            for (node in nodes) {
                if (c == '.') {
                    for (next in node.nexts) {
                        if (next.key != '#') {
                            newNodes.add(next.value)
                        }
                    }
                } else {
                    node.nexts[c]?.let {
                        newNodes.add(it)
                    }
                }
            }
            if (newNodes.isEmpty()) {
                return false
            }
            nodes = newNodes
        }
        return true
    }
}

class TrieNode {
    val nexts = mutableMapOf<Char, TrieNode>()
}
